### PR TITLE
Manual backport: Update go to 1.21.6 and use .go-version file everywhere to release/1.1.x

### PR DESCRIPTION
**Pull request by curtbushko**
Manual backport of #384 to release/1.1.x

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -316,7 +308,7 @@ jobs:
         with:
           name: consul-${{matrix.server.version}}-integration-tests-output
           path: integration-tests/output/
-          if-no-files-found: 'error'
+          if-no-files-found: "error"
```
Comment (zalimeni) [inline .github/workflows/build.yml:311]: Same comment as https://github.com/hashicorp/consul-dataplane/pull/388 re: formatting

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -302,12 +299,7 @@ jobs:
       - uses: actions/checkout@v3
       - uses: actions/setup-go@v3
         with:
-          # pinning this to 1.20.5 because this issue in go-testcontainers occurs
-          # in 1.20.6 with the error "http: invalid Host header, host port waiting failed"
-          # https://github.com/testcontainers/testcontainers-go/issues/1359
-          # remove setting this when the above issue is fixed so that the job will
-          # just get the go version from .go-version.
-          go-version: 1.20.5
+          go-version: ${{ needs.get-go-version.outputs.go-version }}
```
Comment (zalimeni) [inline .github/workflows/build.yml:302]: Missing `needs` here same as https://github.com/hashicorp/consul-dataplane/pull/388 I think

File: .github/workflows/consul-dataplane-checks.yaml [hunk 3]
```diff
@@ -41,16 +30,12 @@ jobs:
       - uses: actions/checkout@v3
       - uses: actions/setup-go@v3
         with:
-          # pinning this to 1.20.5 because this issue in go-testcontainers occurs
-          # in 1.20.6 with the error "http: invalid Host header, host port waiting failed"
-          # https://github.com/testcontainers/testcontainers-go/issues/1359
-          # remove setting this when the above issue is fixed so that the job will
-          # just get the go version from .go-version.
-          go-version: 1.20.5
+          go-version: ${{ needs.get-go-version.outputs.go-version }}
```
Comment (zalimeni) [inline .github/workflows/consul-dataplane-checks.yaml:33]: Missing needs here same as https://github.com/hashicorp/consul-dataplane/pull/388 I think

**Review by zalimeni — APPROVED** [review]
Similar to https://github.com/hashicorp/consul-dataplane/pull/388, LGTM pending a few `needs` and format fixes; approving to unblock after

File: .github/workflows/consul-dataplane-checks.yaml (hunk 3 above)
Comment (curtbushko) [reply .github/workflows/consul-dataplane-checks.yaml:33]: This one has it. The other one did not.

File: .github/workflows/build.yml (hunk 1 above)
Comment (curtbushko) [reply .github/workflows/build.yml]: Fixed!

File: .github/workflows/build.yml (hunk 2 above)
Comment (curtbushko) [reply .github/workflows/build.yml:303]: Fixed!

File: .github/workflows/consul-dataplane-checks.yaml (hunk 3 above)
Comment (zalimeni) [reply .github/workflows/consul-dataplane-checks.yaml:34]: Oops! wrong `integration-tests` 😄

**Pull request merged**